### PR TITLE
feat(commerce): allow multiple recommendations slots at once

### PR DESCRIPTION
**Pull request by Spuffynism**
Two changes are necessary to allow multiple recs slots:
- Passing the `slotId` to the action to fetch recs
- Changing the recommendations state to support multiple recommendations slots

Other changes in the PR follow from these two changes.

With this in place, it'll be easier to complete the work in https://github.com/coveo/ui-kit/pull/3800

[CAPI-762]

[CAPI-762]: https://coveord.atlassian.net/browse/CAPI-762?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/headless/src/features/commerce/recommendations/recommendations-state.ts [hunk 1]
```diff
@@ -10,11 +10,18 @@ export interface RecommendationsState {
...
-  headline: '',
-  error: null,
-  isLoading: false,
-  responseId: '',
-  products: [],
-});
+/**
+ * An object in which each key is a slot identifier, and each value is the corresponding recommendations slice.
+ */
+export type RecommendationsState = Record<string, RecommendationsSlice>;
```
Comment (Spuffynism) [inline packages/headless/src/features/commerce/recommendations/recommendations-state.ts:16]: Here, a `slotId` points to "recommendations", should it point to a "slot" instead?
```suggestion
export type RecommendationsState = Record<string, SlotSlice>;
```

File: packages/headless/src/features/commerce/recommendations/recommendations-state.ts [hunk 2]
```diff
@@ -10,11 +10,18 @@ export interface RecommendationsState {
...
+
+export const getRecommendationsInitialState = (): RecommendationsState => ({});
+export const getRecommendationsSliceInitialState =
+  (): RecommendationsSlice => ({
+    slotId: '',
```
Comment (Spuffynism) [inline packages/headless/src/features/commerce/recommendations/recommendations-state.ts:21]: I'm not sure we need to keep the `slotId` here **and** as state key. I'll see if I can remove it.

File: packages/headless/src/features/commerce/recommendations/recommendations-actions.ts [hunk 3]
```diff
@@ -58,7 +64,7 @@ export interface SlotIdPayload {
   slotId: string;
 }
 
-export const updateRecommendationsSlotId = createAction(
+export const registerRecommendationsSlot = createAction(
   'recommendation/updateSlotId',
```
Comment (BenKandelaars) [inline packages/headless/src/features/commerce/recommendations/recommendations-actions.ts:68]: This name is now out of sync

File: packages/headless/src/features/commerce/recommendations/recommendations-state.ts (hunk 2 above)
Comment (BenKandelaars) [reply packages/headless/src/features/commerce/recommendations/recommendations-state.ts:21]: This was used in the request for recommendations. Now you pass that in the payload, removing its need to be stored in state.

Comment (Spuffynism) [reply packages/headless/src/features/commerce/recommendations/recommendations-state.ts]: Fixed in https://github.com/coveo/ui-kit/pull/3801/commits/5b91a7e2afee7387758c1af891fcbed60c2a4a05

**Review by olamothe — APPROVED** [review]

File: packages/headless/src/features/commerce/recommendations/recommendations-slice.ts [hunk 4]
```diff
@@ -1,31 +1,67 @@
...
+} from './recommendations-state';
 
 export const recommendationsReducer = createReducer(
   getRecommendationsInitialState(),
 
   (builder) => {
     builder
-      .addCase(updateRecommendationsSlotId, (state, action) => {
-        state.slotId = action.payload.slotId;
+      .addCase(registerRecommendationsSlot, (state, action) => {
```
Comment (fbeaudoincoveo) [inline packages/headless/src/features/commerce/recommendations/recommendations-slice.ts:17]: I think we should have a test in `recommendations-slice.test.ts` like:

```typescript
describe('on #registerRecommendationsSlot', () => {
  it('when received #slotId points to an existing recommendations slice in the state, does nothing', () => {
    // check that buildRecommendationsSlice didn't get called
  }
  it('when received #slotId points to no recommendations slice in the state, creates one', () => {
    // check that state[slotId] is equal to the value returned by a mock call to buildRecommendationsSlice()
  }
});
```

File: packages/headless/src/features/commerce/recommendations/recommendations-slice.ts [hunk 5]
```diff
@@ -1,31 +1,67 @@
...
+        }
+
+        state[slotId] = buildRecommendationsSlice();
       })
       .addCase(fetchRecommendations.rejected, (state, action) => {
-        state.error = action.payload ? action.payload : null;
-        state.isLoading = false;
+        const recommendations = state[action.meta.arg.slotId];
+
+        if (!recommendations) {
```
Comment (fbeaudoincoveo) [inline packages/headless/src/features/commerce/recommendations/recommendations-slice.ts:29]: Here we should test that it does not throw when the `slotId` in the action payload corresponds to no recommendations in the state.

Same comment for `fetchRecommendations.fulfilled` and `fetchRecommendations.pending`

**Review by fbeaudoincoveo — APPROVED** [review]

File: packages/headless/src/features/commerce/recommendations/recommendations-state.ts (hunk 1 above)
Comment (Spuffynism) [reply packages/headless/src/features/commerce/recommendations/recommendations-state.ts]: Good point! Applied this in https://github.com/coveo/ui-kit/pull/3801/commits/53b05cca263a9b71e330a2218bbe4e891d16023a

**Review by BenKandelaars — APPROVED** [review]
LGTM

**Pull request merged**